mesonbuild/dependencies/base.py

```diff
@@ -317,11 +317,11 @@ def get_variable(self, *, cmake: T.Optional[str] = None, pkgconfig: T.Optional[s
         raise DependencyException(f'Could not get an internal variable and no default provided for {self!r}')
 
     def generate_link_whole_dependency(self) -> Dependency:
-        from ..build import SharedLibrary
+        from ..build import SharedLibrary, CustomTarget, CustomTargetIndex
         new_dep = copy.deepcopy(self)
-        # XXX: we need to check that CustomTargets are not SharedLibs...
-        if any(isinstance(x, SharedLibrary) for x in new_dep.libraries):
-            raise MesonException(f'Cannot convert a dependency to link_whole with shared libraries')
+        for x in new_dep.libraries:
+            if isinstance(x, SharedLibrary) or (isinstance(x, (CustomTarget, CustomTargetIndex)) and x.links_dynamically()):
+                raise MesonException(f'Cannot convert a dependency to link_whole with shared libraries')
```

Comment:
Can we reorder this slightly and merge this commit with the one two commits before?